File: Algorithm/JewelsandStones.cpp
// // Solution 1: HashTable
// // Time complexity： O(|J| + |S|)
// // Space complexity: O(128) / O(|J|)
// class Solution {
// public:
//     int numJewelsInStones(string J, string S) {
//         std::vector<int> f(128, 0);
//         int ans = 0;
//         for (const char j : J)
//             f[j] = 1;
//         for (const char s : S)
//             ans += f[s] & 1;
//         return ans;
//     }
// };

// v2
class Solution {
public:
    int numJewelsInStones(string J, string S) {
        std::set<char> f(J.begin(), J.end());
        return std::count_if(S.begin(), S.end(),
                             [&f](const char c) { return f.count(c); });
    }
};

File: Algorithm/JewelsandStones.java
// Solution 1: HashTable
// Time complexity: O(|J| + |S|)
// Space complexity: O(128) / O(|J|)
class Solution {
    public int numJewelsInStones(String J, String S) {
        int[] f = new int[128];
        for (final char c : J.toCharArray())
            f[c] = 1;
        int ans = 0;
        for (final char c : S.toCharArray())
            ans += f[c];
        return ans;
    }
}
